### PR TITLE
Replace storage. Move storage-related code from the chain.

### LINKED ISSUE #179: We should not be cloning blocks
Currently `Block` trait is clonable because we clone blocks when we put them in various containers. This is not efficient since blocks are heavy objects.

### LINKED ISSUE #300: use fresh devnet + storage for each integration test run
this is causing issues with inconsistent runs making it hard to track down some problems.. it will also enable us to remove the sequential lock for test runs.

### LINKED ISSUE #391: Integrations tests are interfering with each other
Upon submitting https://github.com/nearprotocol/nearcore/pull/390 I noticed that CI passed on the branch (and locally on my machine), but failed on master.
After examining it closer I noticed that `test_set_get_values_inner` is flaky -- it passes 6 out of 10 times. However, if this test is run separately from other tests `tests_rpc_cli.rs` it passes 20 out of 20 times. Which means tests inside `tests_rpc_cli.rs` interfere with each other.

I am going to disable this test for now until we make sure the tests do not interfere with each other.

### PR DESCRIPTION
**Pull request by nearmax**
* Architecture changes, see justification below:
    * Replaced old storage code with the new one;
    * Moved storage-specific code from chains;
    * Beacon and shard chain now share columns (had to merge the columns because of the bug in kvdb wrapper);
* Integration tests do not interfere anymore;
* Performance optimizations:
    * DevNet performance got visibly faster. No direct measurement, but previously the first transaction issued to DevNet would sometimes take 3-7 secs. Now it is instantaneous;
    * Storage code does not perform unnecessary cloning of the objects. Also the lookup of the existing object does not clone it (if it is in cache), unlike before;
    * Integration tests got 3-4x faster because they run in parallel;

Fixed #391 . Fixed #300. Fixed #179 .

### Architecture changes
https://www.dropbox.com/s/577lq3apprars8f/Screenshot%202019-02-18%2012.03.24.png
The proposed architecture achieves the following two major goals:
1. Separation of storage-specific logic and domain-specific logic;
2. Sequentialization of beacon chain together with beacon storage and shard chain together with shard storage;

Storage-specific logic includes:
* Cache;
* Code that integrates a specific db library. (kvdb wrapper has already proved to be buggy, so there is a good chance that we might decide to swap it for something later);
* Maintaining storage consistency. It is possible that an interrupted client might leave behind an inconsistent storage, e.g.: the blocks column has blocks for which we do not have corresponding headers in the headers column; or also more complex cases with column that represents indices might happen; The code that maintains storage consistency should be isolated;

Domain-specific logic includes:
* Fork-choice rule;
* Keeping track of finalized and pending blocks;
* etc.

There is a very high chance that beacon chain and each of the shard chains are going to be communicated to in parallel. Therefore it is very desirable for them to be sequentialized. However this sequentialization should also include the cache and the storage they are operating on. Therefore it is desirable to have beacon and shard -specific storages that can be sequentialized together with the corresponding chains.

### CODE REVIEW

**Comment by evgenykuzyakov** [comment]
Try rerun benchmarks

**Comment by nearmax** [comment]
> Try rerun benchmarks

@evgenykuzyakov the storage benchmark bypasses most of the storage functionality because it uses the trie only. The real improvement comes in the cache implementation that the benchmark does not use. However, here is the comparison of old vs new results:
Old:
test trie_lookup ... bench:     449,039 ns/iter (+/- 35,359)
test trie_update ... bench:     356,865 ns/iter (+/- 63,417)

New:
test trie_lookup ... bench:     533,775 ns/iter (+/- 37,251)
test trie_update ... bench:     351,627 ns/iter (+/- 44,894)

**Comment by evgenykuzyakov** [comment]
@nearmax I was thinking about wasm benchmarks. But they also don't use storage as much. I probably worth it to create a storage heavy benchmark. Something in order of 250 or 500 reads.

**Comment by nearmax** [comment]
> @nearmax I was thinking about wasm benchmarks. But they also don't use storage as much. I probably worth it to create a storage heavy benchmark. Something in order of 250 or 500 reads.

Yes. I'll do that.

**Comment by evgenykuzyakov** [comment]
Added wasm storage benchmarks in #584

**Comment by nearmax** [comment]
Added benchmark. Divide the following by 10.
```
test storage_save_block     ... bench:     229,673 ns/iter (+/- 80,900)
test storage_save_get_block ... bench:       2,005 ns/iter (+/- 304)
```

File: core/storage/src/trie/mod.rs [hunk 1]
```diff
@@ -854,30 +841,28 @@ mod tests {
...
-            (b"horse".to_vec(), Some(b"stallion".to_vec())),
-            (b"dog".to_vec(), Some(b"puppy".to_vec())),
-            (b"h".to_vec(), Some(b"value".to_vec())),
-        ];
-        let root = test_populate_trie(&trie, &empty_root, changes.clone());
-        let new_root = test_clear_trie(&trie, &root, changes);
-        assert_eq!(new_root, empty_root);
-        assert_eq!(storage.iter(Some(0)).fold(0, |acc, _| acc + 1), 0);
-    }
+//    #[test]
```
Comment (ilblackdragon) [inline core/storage/src/trie/mod.rs]: bring back?

File: node/beacon/src/beacon_chain.rs [hunk 2]
```diff
@@ -26,34 +26,34 @@ impl BeaconBlockChain {
 
 #[cfg(test)]
 mod tests {
-    use std::sync::Arc;
+//    use std::sync::Arc;
```
Comment (ilblackdragon) [inline node/beacon/src/beacon_chain.rs]: remove

File: node/beacon/src/beacon_chain.rs [hunk 3]
```diff
@@ -26,34 +26,34 @@ impl BeaconBlockChain {
...
-            SignedBeaconBlock::new(0, CryptoHash::default(), vec![], CryptoHash::default());
-        let bc = BlockChain::new(genesis.clone(), storage);
-        assert_eq!(bc.get_block(&BlockId::Hash(genesis.block_hash())).unwrap(), genesis);
-        assert_eq!(bc.get_block(&BlockId::Number(0)).unwrap(), genesis);
-    }
+    use primitives::hash::CryptoHash;
+    use storage::test_utils::create_beacon_shard_storages;
+
+//    #[test]
+//    fn test_genesis() {
```
Comment (ilblackdragon) [inline node/beacon/src/beacon_chain.rs]: bring back?

File: core/storage/src/trie/mod.rs (hunk 1 above)
Comment (nearmax) [reply core/storage/src/trie/mod.rs]: Now, I remember why. This unittests verifies some trie stuff by accessing the storage column directly. I have commented out that specific assertion.

File: node/beacon/src/beacon_chain.rs (hunk 2 above)
Comment (nearmax) [reply node/beacon/src/beacon_chain.rs]: I'll keep for the reference and remove after I add finalized, pending block logic.

File: node/beacon/src/beacon_chain.rs (hunk 3 above)
Comment (nearmax) [reply node/beacon/src/beacon_chain.rs]: Done.

**Pull request merged**